runtime/src/main/kotlin/kotlin/native/concurrent/SharedRef.kt

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+
+package kotlin.native.concurrent
+
+import kotlin.native.internal.*
+
+@SymbolName("Kotlin_SharedRef_createSharedRef")
+external private fun createSharedRef(value: Any): NativePtr
+
+/**
+ * A frozen shared reference to a Kotlin object
+ *
+ * Can be safely passed between workers, but can only be accessed on the worker it was created on,
+ * unless the referred object is frozen too
```

Comment:
How is it supposed to be used from workers with this contract? Is there a real-world usage sample somewhere?

Reply:
For example, `(Disposable)SharedRef` is a replacement for `StableRef`, when the latter is (mis)used to share references to non-shared objects. See
https://github.com/Kotlin/kotlinx.coroutines/blob/b2ecddc0cc18dd46cb2e3513841177f2174ca113/kotlinx-coroutines-core/native/src/internal/Sharing.kt#L151

Reply:
This doesn't make it more clear. 
I've noticed I'm not the first one who didn't get it, and you've already explained it once: https://github.com/JetBrains/kotlin-native/pull/4044#discussion_r402245428
Perhaps, it should be explained in the docs of the class.

Reply:
IMO, the name `SharedReference` doesn't convey the concept behind it:
> The purpose is to share reference without sharing contents.

Take the following analogy: when I share a link to some document, everyone who has that link can access that document — this is what can be called a shared reference. That's not the case here, because only one worker can actually access the referenced object.

Reply:
> when I share a link to some document, everyone who has that link can access that document — this is what can be called a shared reference. 

The new reference implemented in this PR is more like a link to a private document which you can save to some public document: if anybody but you follows this link, he/she gets "no access exception".

Do you have better naming suggestion? Maybe @elizarov has some ideas?

Reply:
Possible names:
* `Moniker`
* `IdentityReference`
* `OpaqueReference`
* `ForeignReference`
* `ShieldedReference`
* `InsulatedReference`
* `SafeReference`

Reply:
- `ExclusiveReference`
- `WorkerBoundReference`

Reply:
More:
* `WorkerConfinedReference`
* `WorkerLocalReference`

I like the names that start with `Worker` as they better highlight cases when these classes should be used (that is, when you use workers) and it would be even better if/when they expose `val worker: Worker` property.

Reply:
`WorkerConfinedReference` seems fine to me.

Reply:
Problem is: this name doesn't reflect the fact that this reference itself is shareable.

Reply:
I'll settle on `WorkerBoundReference`, if that's alright.